### PR TITLE
Factor out code storage and add incremental writes and checkpoints

### PR DESCRIPTION
**Pull request by HerbertJordan**
This PR factors out the handling of codes from the MPT state into its own `codes` container and extends this container with the ability to
- incrementally write codes instead of full flushes
- support checkpoints

The former addition helps reducing the shutdown time (see #938) and is a requirement for efficient checkpoint support required by #946.

### CODE REVIEW

**Review by petr-hanzl — APPROVED** [review]
I actually did part of this code base in my PR Im currently working on, so this it removes a little bit of code from it. Thanks!

File: go/database/mpt/codes.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+			return nil, err
+		}
+	}
+
+	data, size, err := readCodesAndSize(file)
+	if err != nil {
+		return nil, err
+	}
+
+	committed := filepath.Join(directory, "committed.json")
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/database/mpt/codes.go]: Done

File: go/database/mpt/codes.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+func (c *codes) GetMemoryFootprint() *common.MemoryFootprint {
+	var sizeCodes uint
+	c.mutex.Lock()
+	for k, v := range c.codes {
+		sizeCodes += uint(len(k) + len(v))
+	}
+	c.mutex.Unlock()
+	return common.NewMemoryFootprint(unsafe.Sizeof(*c) + uintptr(sizeCodes))
```
Comment (petr-hanzl) [inline go/database/mpt/codes.go:142]: I wonder, why `unsafe.Sizeof(*c)` does not include the map, why do you need to count the `sizeCodes` 'by hand'?

**Review by petr-hanzl — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**